src/Symfony/Component/DependencyInjection/Compiler/RegisterEnvProvidersPass.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Argument\ServiceClosureArgument;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\EnvProviderInterface;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\ParameterBag\EnvPlaceholderParameterBag;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Creates the container.env_providers_locator service.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class RegisterEnvProvidersPass implements CompilerPassInterface
+{
+    private static $allowedTypes = array('array', 'bool', 'float', 'int', 'string');
+
+    public function process(ContainerBuilder $container)
+    {
+        $types = array();
+        $providers = array();
+        foreach ($container->findTaggedServiceIds('container.env_provider') as $id => $tags) {
+            foreach ($tags as $attr) {
+                if (!$r = $container->getReflectionClass($class = $container->getDefinition($id)->getClass())) {
+                    throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
+                } elseif (!$r->isSubclassOf(EnvProviderInterface::class)) {
+                    throw new InvalidArgumentException(sprintf('Service "%s" must implement interface "%s".', $id, EnvProviderInterface::class));
+                }
+                foreach ($class::getProvidedTypes() as $prefix => $type) {
+                    $providers[$prefix] = new ServiceClosureArgument(new Reference($id));
+                    $types[$prefix] = self::validateProvidedTypes($type, $class);
+                }
+            }
+        }
+
+        if ($providers) {
+            if ($container->getParameterBag() instanceof EnvPlaceholderParameterBag) {
+                $container->getParameterBag()->setProvidedTypes($types);
+            }
+            $container->register('container.env_providers_locator', ServiceLocator::class)
+                ->setArguments(array($providers))
+            ;
+        }
+    }
+
+    private static function validateProvidedTypes($types, $class)
```

Comment:
What's the benefit of making this function static, as it is already private?

Reply:
none, but another reviewer asked for it, and there is neither no benefit not doing it static :)